proposed/before_bundle_build_hook.md

```diff
@@ -0,0 +1,49 @@
+# Add ability to sign extractable binaries included in a singlefile app.
+
+There is a known issue around singlefile apps when the user needs to sign included artefacts or otherwise modify them. A typical solution for regular build artefacts is to sign binaries after the build is completed.
+A problem arises for files that are included in a singlefile bundle. The fact that included files are not signed is observable for extractable files, but the user cannot easily address that.
+
+Requiring that constituent files are signed prior to building a singlefile app could be highly inconvenient to existing build pipelines.
+It is particularly problematic for a file that exists only in a transient form before its inclusion in a single file bundle. This includes the app main binary itself (as in `hello.dll` for a hello app) and composite r2r image, when composite r2r is used. The build does not currently present a documented opportunity for the user to observe and modify files before inclusion into the bundle.
+
+In addition to signing a similar difficulties may affect other scenarios that rely on post-build modification of build artefacts. For example obfuscation and instrumentation.
+
+## Current state (as of .NET 6)
+
+The problem of signing applies primarily to files that are extracted to the file system as a part of execution. For the assemblies that are not extracted the problem is generally not a concern. 
+With transitioning to superhost technology on all platforms in 6.0 a need to extract files is greatly reduced as native binaries that belong to the dotnet platform are no longer extracted and in fact cannot be extracted, since they are monolithically linked with the app. 
+
+The files that still can be extracted:
+-	Native dependencies included via a nuget references. (this is, technically, the only truly supported scenario to include a native dependency)
```

Comment:
What about P2P references to VCX projects? It's not obvious. The other thing that should be done is tell the user to hook their signing targets to AfterBuild, but that has it's own set of limitations like machine marshalling (which was described in the original issue).

Reply:
Do P2P references to VCX actually work? The typical problem with native dependencies in .NET apps is that the .NET app when consuming the dependency needs to know with RID the dependency supports and potentially consume multiple versions of the native dependency (for multiple RIDs - for portable apps).

I agree that if there's a supported scenario in this area it should be mentioned in the doc though.

Reply:
There's the issue of I am not sure if it's supported, but I know it works and it is used by and large by customers (at least internal ones). They do P2P and then add the dll's needed to Content* ItemGroups.

Reply:
I assume that whatever supported or unsupported mechanism used, ultimately it would need to arrange the inclusion of a native file into the list of bundled files. 

I'd like to not get deep into how an app may gain a native dependency as it is a separate concern and should be orthogonal to this proposal. I just wanted to mention that native dependencies do exist, but most likely signed as a part of their build and prior to building of the app that consumes them. 

In either case the proposed mechanism should be sufficient to sign native dependencies if such need arises.

Reply:
> There's the issue of I am not sure if it's supported, but I know it works and it is used by and large by customers (at least internal ones). They do P2P and then add the dll's needed to Content* ItemGroups.

Assuming this works with singlefile apps, the files must make it to the current `_FilesToBundle` ItemGroup. In such case the proposed solution will work if these files need to be signed in `BeforeBundle`